Functions contain only workers ( without any statements ) fail at Semantic Phase.

```
function forkJoinWithWorkers() {
    worker newWorker1 {
       int a = 5;
    }

    worker newWorker2 {
       int a = 5;
    }
}

```
Ballerina Language parser, we allow above functions with only workers ( making statements optional ), but at Semantic phase this fails giving following exception.

```
Caused by: java.lang.ArrayIndexOutOfBoundsException: -1
	at org.ballerinalang.util.semantics.SemanticAnalyzer.checkAndAddReturnStmt(SemanticAnalyzer.java:4007)
	at org.ballerinalang.util.semantics.SemanticAnalyzer.visit(SemanticAnalyzer.java:662)
	at org.ballerinalang.model.BallerinaFunction.accept(BallerinaFunction.java:230)
	at org.ballerinalang.util.semantics.SemanticAnalyzer.visit(SemanticAnalyzer.java:264)
	at org.ballerinalang.model.BLangPackage.accept(BLangPackage.java:283)
	at org.ballerinalang.util.semantics.SemanticAnalyzer.visit(SemanticAnalyzer.java:207)
	at org.ballerinalang.model.BLangProgram.accept(BLangProgram.java:133)
	at org.ballerinalang.BLangCompiler.compile(BLangCompiler.java:71)
	at org.ballerinalang.nativeimpl.util.BTestUtils.getProgramFile(BTestUtils.java:63)
```

--- a/modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java
+++ b/modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java
@@ -674,12 +674,7 @@ public class SemanticAnalyzer implements NodeVisitor {
 
             BlockStmt blockStmt = function.getCallableUnitBody();
             blockStmt.accept(this);
-
-            if (function.getReturnParameters().length > 0 && !blockStmt.isAlwaysReturns()) {
-                BLangExceptionHelper.throwSemanticError(function, SemanticErrors.MISSING_RETURN_STATEMENT);
-            }
-
-            checkAndAddReturnStmt(function.getReturnParamTypes().length, blockStmt);
+            checkAndAddReturnStmt(function);
         }
 
         resolveWorkerInteractions(function);
@@ -727,12 +722,7 @@ public class SemanticAnalyzer implements NodeVisitor {
 
             BlockStmt blockStmt = action.getCallableUnitBody();
             blockStmt.accept(this);
-
-            if (action.getReturnParameters().length > 0 && !blockStmt.isAlwaysReturns()) {
-                BLangExceptionHelper.throwSemanticError(action, SemanticErrors.MISSING_RETURN_STATEMENT);
-            }
-
-            checkAndAddReturnStmt(action.getReturnParameters().length, blockStmt);
+            checkAndAddReturnStmt(action);
         }
         resolveWorkerInteractions(action);
 
@@ -4097,26 +4087,29 @@ public class SemanticAnalyzer implements NodeVisitor {
     /**
      * Helper method to add return statement if required.
      *
-     * @param returnParamCount No of return parameters.
-     * @param blockStmt        Block statement to which to add the return statement.
+     * @param callableUnit action/function.
      */
-    private void checkAndAddReturnStmt(int returnParamCount, BlockStmt blockStmt) {
-        if (returnParamCount != 0) {
+    private void checkAndAddReturnStmt(CallableUnit callableUnit) {
+        BlockStmt blockStmt = callableUnit.getCallableUnitBody();
+        ParameterDef[] retParams = callableUnit.getReturnParameters();
+        if (retParams.length > 0 && !blockStmt.isAlwaysReturns()) {
+            throw BLangExceptionHelper.getSemanticError(callableUnit.getNodeLocation(),
+                    SemanticErrors.MISSING_RETURN_STATEMENT);
+        } else if (blockStmt.isAlwaysReturns()) {
             return;
         }
 
         Statement[] statements = blockStmt.getStatements();
         int length = statements.length;
-        Statement lastStatement = statements[length - 1];
-        if (!(lastStatement instanceof ReturnStmt)) {
-            NodeLocation blockLocation = blockStmt.getNodeLocation();
-            NodeLocation endOfBlock = new NodeLocation(blockLocation.getPackageDirPath(),
-                    blockLocation.getFileName(), blockLocation.stopLineNumber);
-            ReturnStmt returnStmt = new ReturnStmt(endOfBlock, null, new Expression[0]);
-            statements = Arrays.copyOf(statements, length + 1);
-            statements[length] = returnStmt;
-            blockStmt.setStatements(statements);
-        }
+        NodeLocation blockLocation = blockStmt.getNodeLocation();
+        NodeLocation endOfBlock = new NodeLocation(blockLocation.getPackageDirPath(),
+                blockLocation.getFileName(), blockLocation.stopLineNumber);
+        ReturnStmt returnStmt = new ReturnStmt(endOfBlock, null, new Expression[0]);
+
+        int lengthWithReturn = length + 1;
+        statements = Arrays.copyOf(statements, lengthWithReturn);
+        statements[lengthWithReturn - 1] = returnStmt;
+        blockStmt.setStatements(statements);
     }
 
     private void checkAndAddReplyStmt(BlockStmt blockStmt) {

--- a/modules/ballerina-core/src/test/java/org/ballerinalang/model/functions/FunctionTest.java
+++ b/modules/ballerina-core/src/test/java/org/ballerinalang/model/functions/FunctionTest.java
@@ -1,0 +1,51 @@
+/*
+*  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+*  Unless required by applicable law or agreed to in writing,
+*  software distributed under the License is distributed on an
+*  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+*  KIND, either express or implied.  See the License for the
+*  specific language governing permissions and limitations
+*  under the License.
+*/
+package org.ballerinalang.model.functions;
+
+import org.ballerinalang.core.utils.BTestUtils;
+import org.ballerinalang.model.values.BValue;
+import org.ballerinalang.util.codegen.ProgramFile;
+import org.ballerinalang.util.program.BLangFunctions;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * Test cases for functions.
+ */
+public class FunctionTest {
+
+    private ProgramFile programFile;
+
+    @BeforeClass
+    public void setup() {
+        programFile = BTestUtils.getProgramFile("lang/functions/function-with-no-return-stmt.bal");
+    }
+
+    @Test(description = "Test empty function scenario")
+    public void testEmptyFunction() {
+        BValue[] returns = BLangFunctions.invokeNew(programFile, "emptyFunction");
+        Assert.assertEquals(returns.length, 0);
+    }
+
+    @Test(description = "Test function with empty default worker")
+    public void testFunctionWithEmptyDefaultWorker() {
+        BValue[] returns = BLangFunctions.invokeNew(programFile, "funcEmptyDefaultWorker");
+        Assert.assertEquals(returns.length, 0);
+    }
+}

--- a/modules/ballerina-core/src/test/resources/lang/functions/function-with-no-return-stmt.bal
+++ b/modules/ballerina-core/src/test/resources/lang/functions/function-with-no-return-stmt.bal
@@ -1,0 +1,12 @@
+function emptyFunction() {
+}
+
+function funcEmptyDefaultWorker() {
+    worker newWorker1 {
+        int a = 5;
+    }
+
+    worker newWorker2 {
+        int a = 5;
+    }
+}

